src/lib.rs

```diff
@@ -142,17 +145,24 @@ pub mod doc_test {
     doctest!("../README.md", readme_md);
 }
 
-const EXPECT_INIT: &str = "PyO3 Asyncio has not been initialized";
-
 static ASYNCIO: OnceCell<PyObject> = OnceCell::new();
 static ENSURE_FUTURE: OnceCell<PyObject> = OnceCell::new();
-static EVENT_LOOP: OnceCell<PyObject> = OnceCell::new();
+
+const EXPECT_INIT: &str = "PyO3 Asyncio has not been initialized";
+static CACHED_EVENT_LOOP: OnceCell<PyObject> = OnceCell::new();
 static EXECUTOR: OnceCell<PyObject> = OnceCell::new();
-static CALL_SOON: OnceCell<PyObject> = OnceCell::new();
-static CREATE_FUTURE: OnceCell<PyObject> = OnceCell::new();
 
-fn ensure_future(py: Python) -> &PyAny {
-    ENSURE_FUTURE.get().expect(EXPECT_INIT).as_ref(py)
+fn ensure_future<'p>(py: Python<'p>, awaitable: &'p PyAny) -> PyResult<&'p PyAny> {
+    ENSURE_FUTURE
+        .get_or_try_init(|| -> PyResult<PyObject> {
+            Ok(asyncio(py)?.getattr("ensure_future")?.into())
+        })?
+        .as_ref(py)
+        .call1((awaitable,))
```

Comment:
`ensure_future` either requires a `loop=` kv parameter or must be run on the same thread that the asyncio loop is from, as it calls `get_event_loop()` internally, is this intended? shouldn't we guard against the possibility that this can be called inside a non-python thread?

Here's the source code for `ensure_future`, fwiw;

```py
def ensure_future(coro_or_future, *, loop=None):
    """Wrap a coroutine or an awaitable in a future.

    If the argument is a Future, it is returned directly.
    """
    if coroutines.iscoroutine(coro_or_future):
        if loop is None:
            loop = events.get_event_loop()
        task = loop.create_task(coro_or_future)
        if task._source_traceback:
            del task._source_traceback[-1]
        return task
    elif futures.isfuture(coro_or_future):
        if loop is not None and loop is not futures._get_loop(coro_or_future):
            raise ValueError('loop argument must agree with Future')
        return coro_or_future
    elif inspect.isawaitable(coro_or_future):
        return ensure_future(_wrap_awaitable(coro_or_future), loop=loop)
    else:
        raise TypeError('An asyncio.Future, a coroutine or an awaitable is '
                        'required')
```

Reply:
That's true, but I believe it is being used correctly already. `ensure_future` is only called within the `PyEnsureFuture` callback, which is only called using `call_soon_threadsafe`. `call_soon_threadsafe` will schedule the callback to run on the event loop's thread.